BIP159 updates

jonatack opened the pull request:
A few clarifications following a discussion with BIP author Jonas Schnelli yesterday and then thanks to review feedback from both Jonas and Murch.

These are compatible with the reference implementation in the BIP and with the current implementation in Bitcoin Core.

It is worthwhile to read the discussion in reference implementation https://github.com/bitcoin/bitcoin/pull/10387 (particularly https://github.com/bitcoin/bitcoin/pull/10387#discussion_r156861038).

jonatack commented:
cc @jonasschnelli

jonatack commented:
Did another pass and seems ready for review.

jonasschnelli commented on bip-0159.mediawiki:
```diff
@@ -13,15 +13,16 @@
-# Pruned peers can relay blocks, headers, transactions, addresses and can serve a limited number of historical blocks, thus they should have a way how to announce their service(s)
-# Peers no longer in initial block download should consider connecting some of its outbound connections to pruned peers to allow other peers to bootstrap from non-pruned peers
+# Pruned peers can relay blocks, headers, transactions, and addresses, but they guarantee serving only a minimum limited number of historical blocks; thus, they should have a way to announce their service(s)
+# Pruned peers are defined by this BIP in service bit terms as those having <code>NODE_NETWORK_LIMITED</code> set and any service bit that signals serving the complete block chain, like <code>NODE_NETWORK</code>, not set
```
I think this is better but still hard to read.
Maybe we should describe the expected behavior when both service bits (`NODE_NETWORK` and `NODE_NETWORK_LIMITED`) are set?

jonatack replied:
Good point. Rewrote it, and moved it out of the motivation section into the specification instead.

jonatack commented:
Updated the second commit per review feedback: [1b17eea..b680873](https://github.com/bitcoin/bips/compare/1b17eea05dba7f27c8defd6686658817208fc36d..b68087347add9e412e1e6d5926f2816edf1126ef?short_path=6d04491#diff-6d044914452f6bf5ca7bdb9d3d7065c451a428bea1ffefbc66b47292f3aa587f)

jonatack commented:
(I might propose adding a changelog to this BIP if this moves forward.)

murchandamus commented on bip-0159.mediawiki:
```diff
@@ -34,6 +34,8 @@ This BIP proposes a new service bit
 | NODE_NETWORK_LIMITED || bit 10 (0x400) || If signaled, the peer <I>MUST</I> be capable of serving at least the last 288 blocks (~2 days).
 |}
 
+Pruned/limited peers <I>MUST</I> have <code>NODE_NETWORK_LIMITED</code> set and any service bit that signals serving the complete block chain (e.g. <code>NODE_NETWORK</code>) <I>MUST NOT</I> be set. Signaling serving all historical blocks overrides <code>NODE_NETWORK_LIMITED</code>.
```
Actually, seeing line 34, defining that `NODE_NETWORK_LIMITED` means that a node is capable of serving _at least_ the last 288 blocks makes me think that it would be correct for nodes with the entire blockchain to indicate both `NODE_NETWORK` and `NODE_NETWORK_LIMITED`.

I don’t understand the motivation for the first half-sentence. A pruned node should be permitted to not signal `NODE_NETWORK_LIMITED`. As far as I understand you want to clarify that a pruned node should not be advertising `NODE_NETWORK`?

```suggestion
Pruned/limited peers <I>MUST NOT</I> set a service bit that signals serving the complete block chain (e.g. <code>NODE_NETWORK</code>.
```

jonatack replied:
Done, nice simplification.

murchandamus commented on bip-0159.mediawiki:
```diff
@@ -42,13 +44,15 @@ Full nodes following this BIP <I>SHOULD</I> relay address/services (<code>addr</
-Peers may have different prune depths (depending on the peers configuration, disk space, etc.) which can result in a fingerprinting weakness (finding the prune depth through getdata requests). NODE_NETWORK_LIMITED supporting peers <I>SHOULD</I> avoid leaking the prune depth and therefore not serve blocks deeper than the signaled <code>NODE_NETWORK_LIMITED</code> threshold (288 blocks).
+Peers may have different prune depths (depending on the peers configuration, disk space, etc.), which can result in a fingerprinting weakness (finding the prune depth through getdata requests).
+
+Pruned peers (<code>NODE_NETWORK_LIMITED</code> nodes that do not also signal serving the full block chain) should avoid leaking the prune depth and therefore <I>SHOULD NOT</I> serve blocks deeper than the signaled <code>NODE_NETWORK_LIMITED</code> threshold of 288 blocks.
```
This seems unnecessarily specific. A node that signals `NODE_NETWORK_LIMITED` but not `NODE_NETWORK` should serve only the last 288 blocks, regardless of whether they are pruned or not.

```suggestion
Peers that signal <code>NODE_NETWORK_LIMITED</code> that do not also signal <code>NODE_NETWORK</code> should avoid leaking the prune depth and therefore <I>SHOULD NOT</I> serve blocks deeper than the signaled <code>NODE_NETWORK_LIMITED</code> threshold of 288 blocks.
```

jonatack replied:
Thanks, done.

murchandamus reviewed (commented):
Good clean-up on the text, but I think I understand the service signals differently than you. Given that `NODE_NETWORK_LIMITED` indicates being able to at least serve 288 blocks, it seems perfectly compatible with also signaling `NODE_NETWORK`.

jonatack commented:
Updated to take the feedback from @jonasschnelli and @murchandamus (thanks!) and it looks much better. Please let me know what you think.

murchandamus commented on bip-0159.mediawiki:
```diff
@@ -20,7 +20,7 @@ Define a service bit that allow pruned peers to signal their limited services
 Bitcoin right now only offers the NODE_NETWORK service bit which indicates that a peer can serve
 all historical blocks.
-# Pruned peers can relay blocks, headers, transactions, addresses and can serve a limited number of historical blocks, thus they should have a way how to announce their service(s)
+# Pruned peers can relay blocks, headers, transactions, and addresses, but they guarantee serving only a minimum limited number of historical blocks; thus, they should have a way to announce their service(s)
```
Feel free to ignore, but it would make more sense to me to put the "only" before guarantee, otherwise it sounds like they guarantee not serving more than the minimum, which doesn’t match the specification. Also, isn’t "minimum limited" done twice?

```suggestion
# Pruned peers can relay blocks, headers, transactions, and addresses, but they only guarantee serving a minimum limited number of historical blocks; thus, they should have a way to announce their service(s)
```

jonatack replied:
Thanks, moved "only" as you suggest and dropped "limited".

murchandamus reviewed (commented):
ACK b3e6e24de5cebaa630534e5683c36c4e24831818

Good improvement, one optional nit, if you feel like taking it.

jonasschnelli commented on bip-0159.mediawiki:
```diff
@@ -34,6 +34,8 @@ This BIP proposes a new service bit
 | NODE_NETWORK_LIMITED || bit 10 (0x400) || If signaled, the peer <I>MUST</I> be capable of serving at least the last 288 blocks (~2 days).
 |}
 
+Pruned/limited peers <I>MUST NOT</I> set a service bit that signals serving the complete block chain (e.g., <code>NODE_NETWORK</code>.
```
Should we add an info that "normal" full nodes (`NODE_NETWORK`) can also signal `NODE_NETWORK_LIMITED` as it is inclusive (which would reflect the current behavior on the network)?

jonatack commented:
Thanks @jonasschnelli and @murchandamus for the second round of review. Took your suggestions, added a missing `)` and removed an extra `)` from my last push, and simplified the fingerprinting recommendation to benefit from the latest changes. Let me know what you think.

murchandamus commented on bip-0159.mediawiki:
```diff
@@ -42,13 +44,15 @@ Full nodes following this BIP <I>SHOULD</I> relay address/services (<code>addr</
 Pruned peers following this BIP may consume more outbound bandwidth.
 
-Light clients (and such) who are not checking the <code>nServiceFlags</code> (service bits) from a relayed <code>addr</code>-message may unwillingly connect to a pruned peer and ask for (filtered) blocks at a depth below their pruned depth. Light clients should therefore check the service bits (and eventually connect to peers signaling <code>NODE_NETWORK_LIMITED</code> if they require [filtered]  ...
+Light clients (and such) who are not checking the <code>nServiceFlags</code> (service bits) from a relayed <code>addr</code>-message may unwittingly connect to a pruned peer and ask for (filtered) blocks at a depth below their pruned depth. Light clients should therefore check the service bits (and eventually connect to peers signaling <code>NODE_NETWORK_LIMITED</code> and not also signaling serv ...
```
Sorry for the piecemeal review, but re-reading that paragraph again, I was thinking that it might be good to describe both cases comprehensively:

```suggestion
Light clients (and such) who are not checking the <code>nServiceFlags</code> (service bits) from a relayed <code>addr</code>-message may unwittingly connect to a pruned peer and ask for (filtered) blocks at a depth below the peer’s pruned depth. Light clients should therefore check the service bits and (1) connect to peers signaling <code>NODE_NETWORK_LIMITED</code> that preferably do not also signal serving the full block chain, if they only require [filtered] blocks around the tip, or (2) connect to peers signaling serving the full block chain if they need data older than the latest 288 blocks. Light clients obtaining peer IPs though DNS seed should use the DNS filtering option.
```

murchandamus replied:
Oh, and I noticed that 

- "though" should be "through" 
- once "(filtered)" is in parentheses, but the other time it’s in brackets "[filtered]"

jonatack replied:
Iterative review FTW -- am the sam. Thanks and will look at updating.

jonatack replied:
Great suggestions -- added a commit by you with them.

murchandamus reviewed (commented):
ACK 22f7f0477c825031a8be80589464766cb2eb2559

One optional elaboration

jonasschnelli commented:
ACK d44f70e77a6b477eed8b1d1fa8d95fb6dc1aaa35

murchandamus reviewed (approved):
LGTM, thanks for working on this.
